### ISSUE TITLE
Convergence of the Grammar + Logic + Model

### ISSUE THREAD

**Issue by mttrbrts** [accordproject/cicero#162]
A Cicero template is composed of three things:
- a template grammar, i.e. natural language text with placeholders
- logic, written in Ergo
- a data model, written in CML.

This triple is very powerful, it allows for
- separation of concerns when designing templates, i.e. a lawyer can build a grammar without pollution by the implementation of the logic
- guaranteed error-free execution for well-typed logic, https://docs.accordproject.org/docs/ergo.html#design-choices
- complex user-defined types, see models.accordproject.org
- re-use of individual template components across templates

However, for simple use cases and for those learning the protocol this increases the barrier to entry.

I propose that the model and logic for a template should be optional components. This gives the natural language primacy.  This will also allow the progressive automation of clauses through the following phases.
1. Text only (but identifiable, versioned and hashed as with all templates).
2. Text + model. Placeholders in the grammar have inline type annotations. Custom types are defined in a model file if necessary.
3. Text + model + logic. As in the current protocol specification. Notice that for execution, a model file is required to define the runtime models (e.g. Request / Response / Emit types). The runtime model does not natually fit within the grammar as the template model does.

In step 2 above, the TemplateModel class can be derived from the annotations in the grammar. There is no execution model so there may be no need for a separate model file.

I also propose that a subset of the Ergo language can be defined inline in the grammar.  For example, this would allow arithmetic expressions to be captured alongside the corresponding text.

> Annex B. Acceptance Criteria
> ...
> Shipments that have a temperature or humidity reading outside the agreed range have a price penalty applied calculated using the Formula for Breach Penalty Calculation below. The breach penalty factor to be used is [{breachPenaltyFactor: Double}].
> Formula for Breach Penalty Calculation:
 >   penalty = number of shipment units x difference between sensor reading and agreed range x breach penalty factor
<blockquote>
<pre>
[{ergo:
define function calculatePenalty(units: Double, breachDelta: Double}: Double {
  return units * breachDelta * clause.breachPenaltyFactor
}
}]
</pre>
</blockquote>
For contract templates  that include references to an external clause template in the grammar. The logic and model can be inferred from that relationship. This means that the model and logic do not also need to reference externally defined files.

Collectively, these proposals allow for a spectrum of template complexity. From simple templates defined only by a grammar file with inline type annotations for the template model, through to complex templates with re-useable modules of logic and custom-types.

**Comment by dselman** [comment]
I support this but it should be noted that the easier we make (2) the less likely people are to reuse models across clauses/contracts, which can create the "tower of Babel" problem. Is only suitable for primitive types?

**Comment by mttrbrts** [comment]
I should make my comment clearer.
> Custom types are defined in a model file if necessary

By "if necessary", I mean if custom types are necessary. Complex types should not be defined in-line. However an attribute of the template model may be annotated with the name of a complex type that is defined in the model-file.

**Comment by dselman** [comment]
Doesn’t this mean that type names are now spread across tem and cto files? Eg rename a type in a cto file and you have to potentially check all tem files for refs? What about meta properties like optional, range, arity and decorators for localisation?

It also means that the tem parser now creates a dynamic template model class in the model manager, so you will need a namespace from somewhere.

**Comment by mttrbrts** [comment]
Yes, that's correct. Regarding renaming, we already have that problem between the model and logic files.

I guess we need to decide if the benefit of inline definitions in the grammar outweighs these compromises.

**Comment by jeromesimeon** [comment]
I would go on a limb and suggest that the following distinction might make some sense:
1. (Text + Clause model). Placeholders in the grammar have inline type annotations. Custom types are defined in a model file if necessary. Text only is an example of this where the Text has no parameters and the inferred model is trivially `asset TemplateModel {}`.
2. (Text + Clause model) + (Logic + Runtime Model). As in the current protocol specification. Notice that for execution, a model file is required to define the runtime models (e.g. Request / Response / Emit types). The runtime model does not natually fit within the grammar as the template model does.

**Comment by dselman** [comment]
@jeromesimeon note that it is not desirable to have the request/response types defined in Ergo as they have to be shared with the integration clients that are generating the JSON payloads.
